File: flowjet_valves/public/js/purchase_order.js
frappe.ui.form.on('Purchase Order', {
    refresh(frm) {
        if (frm.doc.docstatus == 1) {
            let items_data = [];

            frm.doc.items.forEach(item => {
                items_data.push({
                    "Item": item.item_code,
                    "Qty": item.qty,
                    "Rate": item.rate,
                    // "Currency": item.currency,
                });
            });

            // Convert array to string for comparison and storage
            let new_details = items_data.map(row => `'${row.Item}': ${row.Qty} qty @ ${row.Rate}`).join(', ');            
            const last_entry = frm.doc.custom_items_history.slice(-1)[0];

            if (!last_entry || last_entry.details !== new_details) {
                frappe.call({
                    method: "flowjet_valves.public.py.purchase_order.update_item_history",
                    args: {
                        po_name: frm.doc.name,
                        items_data: new_details
                    },
                    callback: function (r) {
                        frappe.msgprint(r.message);
                        frm.reload_doc(); // Refresh to show new row
                    }
                });            
            }

            if (
                !["Closed", "Delivered"].includes(frm.doc.status) &&
                frm.has_perm("submit") &&
                (flt(frm.doc.per_billed) < 100 || flt(frm.doc.per_received) < 100)
            ) {
                // Remove the default Close button if it exists
                frm.remove_custom_button("Close", "Status");

                const close_po = (with_new_po = false) => {
                    frappe.call({
                        method: "erpnext.buying.doctype.purchase_order.purchase_order.update_status",
                        args: {
                            status: "Closed",
                            name: frm.doc.name
                        },
                        callback: function () {
                            frm.reload_doc();
                            frappe.show_alert({ message: "Purchase Order closed", indicator: "orange" });

                            if (with_new_po) {
                                frappe.call({
                                    method: "flowjet_valves.public.py.purchase_order.make_new_po_from_remaining",
                                    args: {
                                        source_name: frm.doc.name
                                    },
                                    callback: function (r) {
                                        if (r.message) {
                                            frappe.model.sync(r.message);
                                            frappe.set_route("Form", r.message.doctype, r.message.name);
                                        } else {
                                            frappe.msgprint("No remaining items found for new Purchase Order.");
                                        }
                                    }
                                });
                            }
                        }
                    });
                };

                frm.add_custom_button("Close", () => {
                    if (flt(frm.doc.per_received) < 100) {
                        frappe.confirm(
                            "Do you want to create another Purchase Order for the remaining items?",
                            function () {
                                close_po(true); // Yes: close and make new PO
                            },
                            function () {
                                close_po(false); // No: just close
                            }
                        );
                    } else {
                        close_po(false); // Fully received: just close
                    }
                }, __("Status"));
            }
        }
    },
});

File: flowjet_valves/public/py/purchase_order.py
import frappe
from frappe.model.mapper import get_mapped_doc
from erpnext.buying.doctype.purchase_order.purchase_order import set_missing_values


@frappe.whitelist()
def update_item_history(po_name, items_data):
    doc = frappe.get_doc("Purchase Order", po_name)

    # Allow changes to submitted doc
    doc.ignore_submit = True

    doc.append("custom_items_history", {
        "date": frappe.utils.now_datetime(),
        "user": frappe.session.user,
        "details": items_data
    })

    doc.save()
    return "Items Updated successfully"


@frappe.whitelist()
def make_new_po_from_remaining(source_name):
    def update_item(source_doc, target_doc, source_parent):
        # Only include items with remaining qty
        qty_diff = source_doc.qty - source_doc.received_qty
        if qty_diff > 0:
            target_doc.qty = qty_diff
            target_doc.schedule_date = source_doc.schedule_date
        else:
            target_doc.idx = None  # this will be dropped

    doc = get_mapped_doc(
        "Purchase Order",
        source_name,
        {
            "Purchase Order": {
                "doctype": "Purchase Order",
                "field_map": {
                    "supplier": "supplier",
                    "schedule_date": "schedule_date"
                },
                "validation": {"docstatus": ["=", 1]}
            },
            "Purchase Order Item": {
                "doctype": "Purchase Order Item",
                "field_map": {
                    "name": "prevdoc_detail_docname",
                    "parent": "prevdoc_docname"
                },
                "condition": lambda d: d.qty > d.received_qty,
                "postprocess": update_item,
            }
        },
        target_doc=None,
        postprocess=set_missing_values
    )
    doc.supplier = ''
    doc.supplier_name = ''

    return doc
